rlp/split.go

```diff
@@ -0,0 +1,156 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package rlp
+
+import (
+	"io"
+	"reflect"
+)
+
+// RawValue represents an encoded RLP value and can be used to delay
+// RLP decoding or precompute an encoding. Note that the decoder does
+// not verify whether the content of RawValues is valid RLP.
+type RawValue []byte
+
+var rawValueType = reflect.TypeOf(RawValue{})
+
+// ListSize returns the encoded size of an RLP list with the given
+// content size.
+func ListSize(contentSize uint64) uint64 {
+	return uint64(headsize(contentSize)) + contentSize
+}
+
+// Split returns the content of first RLP value and any
+// bytes after the value as subslices of b.
+func Split(b []byte) (k Kind, content, rest []byte, err error) {
+	k, ts, cs, err := readKind(b)
+	if err != nil {
+		return 0, nil, b, err
+	}
+	return k, b[ts : ts+cs], b[ts+cs:], nil
+}
+
+// SplitString splits b into the content of an RLP string
+// and any remaining bytes after the string.
+func SplitString(b []byte) (content, rest []byte, err error) {
+	k, content, rest, err := Split(b)
+	if err != nil {
+		return nil, b, err
+	}
+	if k == List {
+		return nil, b, ErrExpectedString
+	}
+	return content, rest, nil
+}
+
+// SplitList splits b into the content of a list and any remaining
+// bytes after the list.
+func SplitList(b []byte) (content, rest []byte, err error) {
+	k, content, rest, err := Split(b)
+	if err != nil {
+		return nil, b, err
+	}
+	if k != List {
+		return nil, b, ErrExpectedList
+	}
+	return content, rest, nil
+}
+
+// CountValues counts the number of encoded values in b.
+func CountValues(b []byte) (int, error) {
+	i := 0
+	for ; len(b) > 0; i++ {
+		_, tagsize, size, err := readKind(b)
+		if err != nil {
+			return 0, err
+		}
+		b = b[tagsize+size:]
+	}
+	return i, nil
+}
+
+func readKind(buf []byte) (k Kind, tagsize, contentsize uint64, err error) {
+	if len(buf) == 0 {
+		return 0, 0, 0, io.ErrUnexpectedEOF
+	}
+	b := buf[0]
+	switch {
+	case b < 0x80:
+		k = Byte
+		tagsize = 0
+		contentsize = 1
+	case b < 0xB8:
+		k = String
+		tagsize = 1
+		contentsize = uint64(b - 0x80)
+	case b < 0xC0:
+		k = String
+		tagsize = uint64(b-0xB7) + 1
+		contentsize, err = readSize(buf[1:], b-0xB7)
+	case b < 0xF8:
+		k = List
+		tagsize = 1
+		contentsize = uint64(b - 0xC0)
+	default:
+		k = List
+		tagsize = uint64(b-0xF7) + 1
+		contentsize, err = readSize(buf[1:], b-0xF7)
+	}
+	if err != nil {
+		return 0, 0, 0, err
+	}
+	// Reject values larger than the input slice.
+	if contentsize > uint64(len(buf))-tagsize {
+		return 0, 0, 0, ErrValueTooLarge
+	}
+	// Reject strings that should've been single bytes.
+	if k == String && contentsize == 1 && buf[tagsize] < 128 {
+		return 0, 0, 0, ErrCanonSize
+	}
```

Comment:
The `tagsize` in this condition took me a while to verify that it cannot screw up. Maybe we could simply write `buf[1]` there? If it's a string, it means the `tag == 1` if length <= 55, or tag == [2-9] if length > 55. Since the content check already made sure we're in the first case, tag must be `== 1`. It would seem cleaner to write that there, instead of the variable. But this is really nitpicking, I can live without it :D